### DISCUSSION TITLE
[Compilers] Test coverage

### DISCUSSION THREAD

**Discussion by RomanYarinski** [zkSync-Community-Hub/zksync-developers#152] [Compilers] [answered]
### Which compiler are you using?

- [X] zksolc
- [ ] zkvyper

### Which version of Solidity or Vyper?

v0.8.19

### zksolc/zkvyper version?

v1.3.14

### Explain the issue

```bash
Have a question about testing contracts locally with zkSync:
Failed to run coverage. Is there any way we can check coverage tests if we are working with zkSync ?

System:
  Host: IdeaPad-3-15ALC6 Kernel: 6.2.0-31-generic x86_64 bits: 64
  Desktop: GNOME 42.9 Distro: Ubuntu 22.04.3 LTS (Jammy Jellyfish)


Error massage:

Error in plugin solidity-coverage: Error: Command failed: /home/roman/.cache/hardhat-nodejs/compilers-v2/zksolc/zksolc-v1.3.14 --standard-json  --system-mode  --solc /home/roman/.cache/hardhat-nodejs/compilers-v2/linux-amd64/solc-linux-amd64-v0.8.19+commit.7dd6d404
Contract `contracts/IsolatedWalletFactory.sol:IsolatedWalletFactory` compiling error: The contract `contracts/IsolatedWalletFactory.sol:IsolatedWalletFactory` LLVM IR generator definition pass error: Dependency contract `IsolatedWalletLogic_6059` compiling error: The contract `contracts/IsolatedWalletLogic.sol:IsolatedWalletLogic` LLVM IR generator definition pass error: Library `contracts/lib/RevertReasonParser.sol:RevertReasonParser` not found in the projectThe contract `contracts/IsolatedWalletFactory.sol:IsolatedWalletFactory` LLVM IR generator definition pass error: Dependency contract `IsolatedWalletLogic_6059` compiling error: The contract `contracts/IsolatedWalletLogic.sol:IsolatedWalletLogic` LLVM IR generator definition pass error: Library `contracts/lib/RevertReasonParser.sol:RevertReasonParser` not found in the project
```

This problem does not occur during usual compilation or tests.
```


### Link to your repo?

_No response_

**Comment by RomanYarinski** [comment]
To run the coverage I was helped by installing the libraries:
`NODE_ENV=test yarn hardhat deploy-zksync:libraries --network zkSyncTestnet --private-key 0x7726827caac94a7f9e1b160f7ea819f172f7b6f9d2a97f992c38edeab82d4110 --exported-config-object configObject`

But still I got an unexpected result, as I have **209 successful tests** for different contracts and functions, and as a result I get **0% of coverage** in the report.
![image](https://github.com/zkSync-Community-Hub/zkync-developers/assets/139853615/1819353f-428e-47b5-9d65-eec85ed88d73)

**Reply by MexicanAce** [reply]
Hi @RomanYarinski ,

Could you please post a link to your code? It's possible there's a misconfiguration of your test files.

**Reply by RomanYarinski** [reply]
I can provide my config file.
```
import "@nomicfoundation/hardhat-toolbox";

import "tsconfig-paths/register";
import "hardhat-gas-reporter";
import "solidity-coverage";
import "hardhat-watcher";
import "hardhat-deploy";
import "solidity-docgen";
import "hardhat-dependency-compiler";

// zkSync plugins
import "@matterlabs/hardhat-zksync-deploy";
import "@matterlabs/hardhat-zksync-solc";
import "@matterlabs/hardhat-zksync-verify";
import "@matterlabs/hardhat-zksync-chai-matchers";
import "./tasks/index";

import { HardhatUserConfig } from "hardhat/config";
import envConfig from "./config";

const { DEPLOYER_KEY, INFURA_KEY, ETHERSCAN_API_KEY, POLYGONSCAN_API_KEY, BSCSCAN_API_KEY } = envConfig;

function typedNamedAccounts<T>(namedAccounts: { [key in string]: T }) {
  return namedAccounts;
}

const zkSyncTestnet =
  process.env.NODE_ENV == "test"
    ? {
        url: "http://localhost:3050",
        ethNetwork: "http://localhost:8545",
        zksync: true,
      }
    : {
        url: "https://zksync2-testnet.zksync.dev",
        ethNetwork: "goerli",
        zksync: true,
        verifyURL: "https://zksync2-testnet-explorer.zksync.dev/contract_verification",
      };

const config: HardhatUserConfig = {
  zksolc: {
    version: "latest",
    settings: {
      isSystem: true, 
      libraries: {
        "contracts/lib/RevertReasonParser.sol": {
          RevertReasonParser: "0x76e893827c89B1e2F3EC5225fdA9f90eaD683FD4",
        },
      },
    },
  },
  dependencyCompiler: {
    paths: ["contracts/lib/RevertReasonParser.sol"],
  },
  defaultNetwork: "zkSyncTestnet",
  solidity: {
    version: "0.8.19",
    settings: {
      optimizer: {
        enabled: true,
        runs: 200,
      },
    },
  },
  typechain: {
    outDir: "types/typechain-types",
  },
  networks: {
    hardhat: {
      gasPrice: "auto",
      loggingEnabled: false,
      forking: {
        url: "",
        enabled: false,
      },
      zksync: false,
    },
    zkSyncTestnet,
    zkSyncMainnet: {
      url: "https://zksync2-mainnet.zksync.io",
      ethNetwork: "mainnet", 
      zksync: true,
      verifyURL: "https://zksync2-mainnet-explorer.zksync.io/contract_verification", 
    },
    localhost: {
      url: "http://127.0.0.1:8545",
      zksync: false,
    },
    mainnet: {
      url: `https://mainnet.infura.io/v3/${INFURA_KEY}`,
      chainId: 1,
      accounts: [DEPLOYER_KEY || ""],
      zksync: false,
    },
    goerli: {
      url: `https://goerli.infura.io/v3/${INFURA_KEY}`,
      chainId: 5,
      accounts: [DEPLOYER_KEY || ""],
      zksync: false,
    },
  },
  etherscan: {
    apiKey: {
      mainnet: ETHERSCAN_API_KEY || "",
      goerli: ETHERSCAN_API_KEY || "",
    },
  },
  namedAccounts: typedNamedAccounts({
    deployer: 0,
  }),
  docgen: {
    exclude: ["./mocks"],
    pages: "single",
  },
  watcher: {
    test: {
      tasks: [{ command: "test", params: { testFiles: ["{path}"] } }],
      files: ["./test/**/*"],
      verbose: true,
    },
  },
  gasReporter: {
    enabled: false,
    coinmarketcap: "",
    currency: "USD",
    token: "ETH",
    gasPrice: 28,
  },
};

export default config;
```

I am running tests using the command `NODE_ENV=test yarn hardhat test --network zkSyncTestnet` or `NODE_ENV=test yarn hardhat test` and everything runs fine.

If I am running coverage using the command `NODE_ENV=test yarn hardhat coverage --network zkSyncTestnet` and it does not work for me.
Error message:
```
Error in plugin solidity-coverage: Error: Cannot find module 'ganache-cli'

Require stack:
- /home/roman/work/4ire/vest/contracts-v3-hardhat/node_modules/solidity-coverage/plugins/hardhat.plugin.js
- /home/roman/work/4ire/vest/contracts-v3-hardhat/node_modules/solidity-coverage/plugins/nomiclabs.plugin.js
- /home/roman/work/4ire/vest/contracts-v3-hardhat/node_modules/@nomicfoundation/hardhat-toolbox/dist/src/index.js
- /home/roman/work/4ire/vest/contracts-v3-hardhat/hardhat.config.ts
- /home/roman/work/4ire/vest/contracts-v3-hardhat/node_modules/hardhat/internal/core/config/config-loading.js
- /home/roman/work/4ire/vest/contracts-v3-hardhat/node_modules/hardhat/internal/cli/cli.js
```
If I am running coverage using the command `NODE_ENV=test yarn hardhat coverage` (without specify network) it works, but I have empty report, like describe earlier.

**Reply by MexicanAce** [reply, marked as answer]
We currently don't have support for `solidity-coverage`. But nonetheless, we will investigate the issue as well as find what should be our next steps when it comes to this plugin.

Edit: As a follow-up, we will be allocating time in the near future for addressing support for coverage. We'll post an update when this effort is complete.

**Comment by jakeaukuso** [comment]
Unfortunately, the current version of Solidity coverage tools, such as solidity-coverage, may not fully support zkSync contracts due to the differences in the execution environment and the limited support for zkSync-specific features.

As a workaround, you can try the following approaches:

Simplify your contracts: Remove any zkSync-specific features or dependencies that may be causing compilation errors. This may involve refactoring your code to reduce complexity and remove dependencies on external libraries or contracts.

Manually test your contracts: Instead of relying on automated coverage tools, you can manually test your contracts using traditional testing frameworks like Hardhat or Truffle. Write comprehensive test cases that cover the critical functionality of your contracts and ensure they behave as expected.

Use zkSync-specific testing tools: Explore zkSync-specific testing tools or frameworks that are designed to work with zkSync contracts. These tools may provide better support for zkSync-specific features and help you test your contracts more effectively.

Monitor updates: Keep an eye on updates to Solidity coverage tools and zkSync development. As both ecosystems evolve, new tools and features may become available that improve support for testing zkSync contracts.

Overall, while testing zkSync contracts with existing coverage tools may be challenging, you can still ensure the reliability and correctness of your contracts through thorough manual testing and diligent monitoring of updates and advancements in the ecosystem.

**Comment by kakarukjoseph** [comment]
It seems like there's an issue with the solidity-coverage plugin when attempting to run coverage tests for contracts that depend on other contracts or libraries that are not found in the project. This might be due to the way solidity-coverage interacts with zkSync or how it handles contract dependencies. 

To address this, you might need to ensure that all required contracts and libraries are included in your project or configured properly for the coverage tests. Additionally, consider checking the documentation or seeking support from the solidity-coverage plugin or zkSync community for further assistance in resolving this issue.

**Comment by tumb2ibonsa** [comment]
To test contracts locally with zkSync and check coverage, you need to ensure that all dependencies, including libraries, are properly included and accessible in your project. The error message indicates a compilation issue related to a missing library (`RevertReasonParser.sol`). Ensure that all required dependencies are correctly imported and available for compilation.

**Comment by anggabarber** [comment]
It seems like there is an issue with running coverage tests locally with zkSync. The error message indicates that there is a problem with a dependency contract, specifically the IsolatedWalletLogic contract and its dependency on the RevertReasonParser library. It appears that the library is not found in the project. Without further information or access to the code repository, it is difficult to diagnose the exact issue.
